### ISSUE TITLE
Web interface

### ISSUE THREAD

**Issue by mdwn** [mdwn/mtrack#12]
A web interface should be introduced so that `mtrack` can be monitored and controlled over a network connection.

**Comment by mdwn** [comment]
I'd actually mark this as the next most important thing. In the case of on stage disaster, i.e. MIDI controller fails, the current way of triggering otherwise is the command line.

This is a terrible user experience, so adding in a web server is critical.

**Comment by gulbrand** [comment]
I'd be interested in helping with this.

I could probably put together an Axum-based http server using htmx that has the buttons to control various events. That seems like a good enough incremental start, but I wanted to also get a sense for direction you had in mind. Could go with Leptos for a more flexible evolution. Htmx can go a long ways, but depending on complexity of features and real-timeliness in the browser side, something with closer integration with wasm in the browser might be more appropriate. Do you have a preference?

I can do either.

Also, a bit about me:

For live performance, I currently use Reaper on a Raspberry Pi, a MIDI Captain Foot Controller, an X32 Rack, a Scarlett 18i20 from the pi to the X32. I have a Dante card in the X32, but I don't use Dante live....yet. This works well, but I want a simpler setup and one that is more flexible. Google search led me here. I have a few other interfaces as well and can also help test those out separately from this issue.

It seems like maybe my goals align with the goals of this project? Would love to chat more on discord if you're open to live discussion.

And one caveat:

I need to check with my employer to get approval to contribute to open source (mostly to get a conflict of interest review), I don't see any issues surfacing there. In the meantime, here are a few thoughts:


Regards,

Mark

**Comment by mdwn** [comment]
@gulbrand That would be amazing! I don't have strong opinions on this -- full disclosure, I'm neither well versed in front end stuff nor all that much in Rust. :-D So I'll defer to you here if you'd like to add it.

It does indeed sound like your goals align with this! FWIW, I've more recently tested this with an X32 Rack and the player works well with it.

Not a problem on the employer review.

**Comment by gulbrand** [comment]
I just got the approval to contribute.

Quick question, what do you think about me adding an OSC compatible network server? Then, a separate UI that uses the OSC interface?

**Comment by mdwn** [comment]
Sounds brilliant to me!

**Comment by tobiaslabu** [comment]
I have been experimenting with an OSC integration for mtrack. I need to do some cleanup, but could provide some working code. 

Edit: I've uploaded my take  at implementing OSC in mtrack here: https://github.com/tobiaslabu/mtrack/tree/addOscSupport
It still needs some tests. 
I can relate to the general disclaimer in the README of this project as I am also not very experienced using Rust. I still hope my efforts can help the project.

**Comment by gulbrand** [comment]
@tobiaslabu I took a quick scan and this looks pretty close to what I had in mind. I think we're on the same page generally.

I'm happy to do a deeper review but it looks good to me high level and it might be a week or so before I can circle back to this.

One thing I noticed, though I'm not sure yet if this might be an issue here. Using std::sync::Mutex in the same context as tokio can cause a deadlock because std::sync::Mutex::lock() blocks. If the mutex is used in any async code, trying to lock it will block the runtime and could cause a deadlock. The solution in some cases is to use tokio::sync::Mutex, which doesn't block, but has to be used from an async environment. This is tricky.

I didn't look closely enough at this. Thought I'd point it out just in case.

**Comment by tobiaslabu** [comment]
@gulbrand Thanks for the feedback! The std::sync::Mutex does not seem to pose an issue as of now, but I'll keep it in mind and have a look if replacing it with a Mutex from tokio works in this case. 

I just committed some changes that include tests similar to the MIDI controller test and the functionality seems to work fine so far. I think that a lot can be done in a cleaner way, however, I am not sure if I will get a lot done before January.

**Comment by mdwn** [comment]
Hey all, thought I'd update here. I've been in a coding mood lately and I've added a gRPC server to mtrack: https://github.com/mdwn/mtrack/pull/69.

With this we might be able to lean on something like [gRPCui](https://github.com/fullstorydev/grpcui) for our frontend rather than needing to build something bespoke. What do you all think?

**Comment by tobiaslabu** [comment]
I still think that OSC support would be a great feature. I stumbled on this project looking for a replacement for my band's setup that used a Notebook running Reaper and a Streamdeck mini to send OSC commands to control Reaper. My plan is to use a Raspberry Pi running mtrack instead of the Notebook. 
Also apps like TouchOSC could quickly be set up to control mtrack from a moblile device. 

Maybe both solutions can be used and be configurable similar to the controllers (keyboard, midi). I think we could define a trait for ways to show mtrack's status that could be implemented for different protocols (midi, OSC, gRPC).

**Comment by mdwn** [comment]
Alright: https://github.com/mdwn/mtrack/pull/76 this is checked in now. I had to do a good bit of finagling and generalized your multi approach, @tobiaslabu. Sorry if I ripped too much out.

Do you mind taking a look and trying this out? I tested locally using https://github.com/madskjeldgaard/osccli, which seemed to work.

**Comment by tobiaslabu** [comment]
Thanks a lot! I tried it out with a web interface that I am working on and it works well. Now only  the status commands (get playlist, get play mode, get current song) need to be implemented, right?

**Comment by mdwn** [comment]
That's right! I'm working on that currently.

**Comment by mdwn** [comment]
https://github.com/mdwn/mtrack/pull/77 This PR is in. I did a good amount of refactoring here, so apologies for the renamed OSC bits. My big hold up was trying to get UDP multicasting working with TouchOSC. Have you messed much with this? I failed at figuring this out, so right now there's a fixed list of clients in the mtrack configuration where status is broadcasted to.

**Comment by mdwn** [comment]
One last thing. I've added in a TouchOSC layout [here](https://github.com/mdwn/mtrack/tree/main/touchosc).

I wish I had realized how perfect OSC was for this up front -- this really is a fantastic fit for mtrack.

**Comment by tobiaslabu** [comment]
I haven't figured out setting up multicast yet either. My workaround so far was to manage a list of known endpoints at runtime that received status messages by unicast messages. Whenever an unknown source called a OSC command it would be added to the list. 
I think the current solution is fine for now and I could give multicast another shot once my custom remote is out :).

**Comment by gulbrand** [comment]
I will try to catch up this week and see where I can help. Thanks for pushing this forward so much. OSC definitely felt like the way to go. Glad to see that its working.

**Comment by gulbrand** [comment]
I'm really happy with TouchOSC and the current OSC implementation. Is there still a place for a web interface?

What I mean is I am more inclined to use the OSC and TouchOSC, so I feel less compelled to invest in a web interface, but if that would help someone else, I could still do that.

**Comment by mdwn** [comment]
I'm inclined to agree. I think with OSC in place the web interface becomes less necessary. @tobiaslabu, are you working on an OSC based front end, or did I misread earlier?

If we're all in agreement here, I'm happy to close this.

**Comment by tobiaslabu** [comment]
You're right, I'm working on an OSC based front end. It's also true that the functionality is covered by TouchOSC, but I hope that one day the mentioned app can do a few things more, like managing songs and playlists. 
In my opinion this issue can be closed, the front end will soon get its own repository. 

Edit: I just published the initial version of the web application: https://github.com/tobiaslabu/mtrack-remote/.

**Comment by mdwn** [comment]
I just noticed that you've got the initial version of this out. :-) I'm going to close this. I can't wait to test this out!
